### PR TITLE
Core review breakpoint

### PR DESCRIPTION
**Pull request by SirMorek**
Almost complete, I think I'm just missing packaging and deployment.

### CODE REVIEW

File: matchmaker/player.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+'''
+Abstract Player.
+
+Could have any set of skills, but we only need hockey players for now.
+'''
+
+
+def get_skills(player_json):
+    return {
```
Comment (gralamin) [inline matchmaker/player.py]: Curly braces aren't required, Round brackets `()` would be fine.

File: matchmaker/player.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+'''
+Abstract Player.
```
Comment (gralamin) [inline matchmaker/player.py]: Docstring saids Abstract here, but the Player class isn't using abc. Just say Base class?

File: matchmaker/player.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    def __eq__(self, other):
+        return self.skill_total == other.skill_total
+
+    def __lt__(self, other):
+        return self.skill_total < other.skill_total
+
+
+class HockeyPlayer(Player):
+    SKILLS = ['shooting', 'skating', 'checking']
```
Comment (gralamin) [inline matchmaker/player.py]: If you are using python 3, consider using an enum.

File: matchmaker/assign.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+This module is responsible for providing the ability to assign players to teams.
+
+Assigners are expected to implement different assignment strategies, potentially
+creating different team compositions.
+'''
+
+
+class RoundRobinAssigner(object):
+    '''
+    A basic team builder that simply round-robins through the provided players
```
Comment (gralamin) [inline matchmaker/assign.py]: This doesn't look like its pep257 compliant.

File: matchmaker/assign.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+'''
+This module is responsible for providing the ability to assign players to teams.
+
+Assigners are expected to implement different assignment strategies, potentially
+creating different team compositions.
+'''
+
+
+class RoundRobinAssigner(object):
```
Comment (gralamin) [inline matchmaker/assign.py]: There currently isn't any reason for this to be a class, maybe future commits will make a reason, but unless you want to enforce the same interface on multiple things (which I would then expect an abstract base class), its not a great sign.

File: matchmaker/team.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+'''
+Simple structures for tracking teams.
+
+Expected to have some basic ability to make it easy to add, remove and try out
+players while sorting out players.
+'''
+
+
+class Team(object):
```
Comment (gralamin) [inline matchmaker/team.py]: You added Teams in this commit, but didn't add any tests for them.

File: matchmaker/main.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+'''
+The matchmaker service is responsible for taking the set of players and
```
Comment (gralamin) [inline matchmaker/main.py]: pep257 note again.

File: matchmaker/unittest/test_main.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from matchmaker.main import matchmake
+
+
+def test_matchmake():
+    import json
+    with open("../players.json") as player_file:
```
Comment (gralamin) [inline matchmaker/unittest/test_main.py]: Generally you don't want to open non-test files in unit tests.

File: web-service/matchmaker [hunk 9]
```diff
@@ -0,0 +1 @@
+../matchmaker/
```
Comment (gralamin) [inline web-service/matchmaker]: Such a dirty hack. Hoping this gets fixed later.

File: web-service/app.py [hunk 10]
```diff
@@ -10,6 +12,36 @@
 def root():
     with open("players.json") as player_file:
         player_json = json.loads(player_file.read())
-        teams, remainder = matchmake(player_json['players'], 6)
+        teams, players = matchmake(player_json['players'], 6)
 
-    return "Teams: %s\nRemainder: %s" % (teams, remainder)
+    with open("persistence.json", "w") as outfile:
```
Comment (gralamin) [inline web-service/app.py]: Consider separating some of this logic into its own function to enable easy testing without worrying about the Flask part.

File: web-service/static/style.css [hunk 11]
```diff
@@ -0,0 +1,134 @@
+html {
```
Comment (gralamin) [inline web-service/static/style.css:1]: This is a sort of common thing - treating the UI related changes as not worth breaking apart. I suggest talking in the commit message a bit about why you choose some of the styling, even if it works out to something like "stole bootstrap.css".

File: web-service/templates/base.html [hunk 12]
```diff
@@ -1,20 +1,58 @@
...
+
 <!DOCTYPE html>
 <html lang="en">
   <head>
     <title>Squad Matchmaker</title>
     <link href="{{ url_for('static', filename='style.css') }}" rel="stylesheet">
   </head>
   <body>
+    <style>
+      table, th, td {
```
Comment (gralamin) [inline web-service/templates/base.html]: You have a css document, why put this stuff here instead of applying it to a class or something there?

File: web-service/app.py [hunk 13]
```diff
@@ -1,33 +1,19 @@
 import json
-from flask import Flask, render_template
+from os import remove
```
Comment (gralamin) [inline web-service/app.py]: This commit ("Webservice: Refactor app to add generate/reset buttons" is doing a bunch of unrelated refactoring work (making a PERSISTENCE_FILE variable for example). Consider splitting up.

File: matchmaker/assign.py [hunk 14]
```diff
@@ -4,6 +4,17 @@
 Assigners are expected to implement different assignment strategies, potentially
 creating different team compositions.
 '''
+from copy import deepcopy
+
+
+def evaluate_SSE(teams):
```
Comment (gralamin) [inline matchmaker/assign.py]: Isn't clear what SSE means here.

**Review by gralamin — COMMENTED** [review]
You mentioned you didn't spend time for packaging and deployment ahead of time. I think you would have benefited quite a bit for setting up your environment to deploy any time, with scripts for flake8, flake8-docstrings, etc.

That said, my first round of comments aren't too bad, and some of them are probably ignorable. The design as a whole looks alright - it could be improved, but it seems like it would work. The algorithms choosen appear to be tested in such a way that you understand the consequences of your algorithm,.

File: web-service/templates/base.html (hunk 12 above)
Comment (SirMorek) [reply web-service/templates/base.html]: Legit tried and it refused to work. I don't understand why doing it here worked and doing the same stuff in the stylesheet didn't, so I cut my losses with it and moved on to other parts of the problem.

File: matchmaker/assign.py (hunk 14 above)
Comment (SirMorek) [reply matchmaker/assign.py]: Agreed, worth including a docstring. (Sum of squared errors)

File: matchmaker/unittest/test_main.py (hunk 8 above)
Comment (SirMorek) [reply matchmaker/unittest/test_main.py]: Yeah I should replace that dependency with a new one generated just for testing...

File: matchmaker/assign.py (hunk 5 above)
Comment (SirMorek) [reply matchmaker/assign.py]: Fair, just yielding the methods would have the same outcome. You see here the same problem I had with player where I started the project thinking I wanted ABC patterns but didn't follow through.

File: matchmaker/player.py (hunk 3 above)
Comment (SirMorek) [reply matchmaker/player.py]: Should just delete it, its not used or useful. An enum is probably better for what it was supposed to do, you are correct.

File: matchmaker/player.py (hunk 1 above)
Comment (SirMorek) [reply matchmaker/player.py]: I wanted a dict comprehension, i believe that you are talking about something else...

File: matchmaker/assign.py (hunk 4 above)
Comment (SirMorek) [reply matchmaker/assign.py]: Could you be more specific about what is in violation here? Reading through pep257 the main thing it would be unhappy with is my use of ''' instead of """

**Comment by SirMorek** [comment]
Making a new PR to show the latest tip.
